src/wpublish/QueriesAndChunks.cc

```diff
@@ -489,105 +499,168 @@ QueriesAndChunks::ScanTableSumsMap QueriesAndChunks::_calcScanTableSums() {
-        }
-    } else {
-        // Disabled as too aggressive and vulnerable to bad statistics. DM-11526
-        if (false && uq->_tasksBooted > _maxTasksBooted) {
-            LOGS(_log, LOG_LVL_INFO,
-                 "entire UserQuery booting from " << sched->getName() << " tasksBooted=" << uq->_tasksBooted
-                                                  << " maxTasksBooted=" << _maxTasksBooted);
-            uq->_queryBooted = true;
-            bSched->moveUserQueryToSnail(uq->queryId, sched);
-        }
-    }
-}
+    // Track the total number of Tasks that are running while booted
+    // and move the worst offenders to the snailScan. The worst offender is the UserQuery with
+    // the most `Task`s running that are taking too long.
+    // The following function only counts booted `Task`s that are still EXECUTING,
+    // and `countQIdVect` is order by worst offender to least worst offender.
+    auto [bootedTaskCount, countQIdVect] = _bootedTaskTracker.getTotalBootedTaskCount();
+    // Sum of Tasks in
+    int uqBootedTasksSum = 0;
+    int maxDarkQueries = 5 * _maxTasksBooted;
+    if (bootedTaskCount > maxDarkQueries) {
+        for (auto const& countQId : countQIdVect) {
+            // If enough UserQueries have been booted to get under the Task limit,
+            // stop booting UserQueries.
+            bool enoughTasksBooted = (bootedTaskCount - uqBootedTasksSum) <= maxDarkQueries;
+            LOGS(_log, LOG_LVL_DEBUG,
+                 __func__ << " a uq=" << countQId.qId << " enough=" << enoughTasksBooted
+                          << " bootedTaskCount=" << bootedTaskCount
+                          << " uqBootedTasksSum=" << uqBootedTasksSum
+                          << " sub=" << (bootedTaskCount - uqBootedTasksSum) << " max=" << maxDarkQueries);
+            if (enoughTasksBooted) {
+                break;
+            }
 
-/// Add a Task to the user query statistics.
-void QueryStatistics::addTask(wbase::Task::Ptr const& task) {
-    lock_guard<mutex> guard(_qStatsMtx);
-    if (queryId != task->getQueryId()) {
-        string const msg = "QueryStatistics::" + string(__func__) +
-                           ": the task has queryId=" + to_string(task->getQueryId()) +
-                           " where queryId=" + to_string(queryId) + " was expected.";
-        throw util::Bug(ERR_LOC, msg);
+            // Get query info
+            QueryId qIdToBoot = countQId.qId;
+            QueryStatistics::Ptr queryToBoot = getStats(qIdToBoot);
+            if (queryToBoot != nullptr) {
+                if (_bootUserQuery(queryToBoot, bSched)) {
+                    uqBootedTasksSum += countQId.count;
+                    LOGS(_log, LOG_LVL_DEBUG,
+                         __func__ << " a booted uq=" << countQId.qId << ":count=" << countQId.count
+                                  << " uqBootedTasksSum=" << uqBootedTasksSum);
+                }
+            } else {
+                // This really should never happen.
```

Comment:
Then the code should throw:
```
util::Bug(ERR_LOC, "This really should never happen");
```

Reply:
It shouldn't happen, but if it does, it probably shouldn't crash the worker. I could see this possibly happening if a query got cancelled, but had long running tasks. Not being able to boot a cancelled task should not have any effect on performance.